src/Cache.php

```diff
@@ -50,24 +50,49 @@ public function removeLegacyTags(array $data): array
         if (!$this->symfonyConstraints || !isset($data['packages']['symfony/symfony'])) {
             return $data;
         }
-        $symfonyVersions = $data['packages']['symfony/symfony'];
 
-        foreach ($data['packages'] as $name => $versions) {
-            foreach ($versions as $version => $package) {
-                if ('symfony/symfony' !== $name && 'self.version' !== ($symfonyVersions[preg_replace('/^(\d++\.\d++)\..*/', '$1.x-dev', $version)]['replace'][$name] ?? null)) {
-                    continue;
-                }
-                $normalizedVersion = $package['extra']['branch-alias'][$version] ?? null;
-                $normalizedVersion = $normalizedVersion ? $this->versionParser->normalize($normalizedVersion) : $package['version_normalized'];
-                $provider = new Constraint('==', $normalizedVersion);
-
-                if (!$this->symfonyConstraints->matches($provider)) {
-                    if ($this->io) {
-                        $this->io->writeError(sprintf('<info>Restricting packages listed in "symfony/symfony" to "%s"</info>', $this->symfonyRequire));
-                        $this->io = null;
-                    }
-                    unset($data['packages'][$name][$version]);
+        $symfonyPackages = [];
+        $symfonySymfony = $data['packages']['symfony/symfony'];
+
+        foreach ($symfonySymfony as $version => $composerJson) {
+            if ('dev-master' === $version) {
+                $normalizedVersion = $this->versionParser->normalize($composerJson['extra']['branch-alias']['dev-master']);
+            } else {
+                $normalizedVersion = $composerJson['version_normalized'];
+            }
+
+            if ($this->symfonyConstraints->matches(new Constraint('==', $normalizedVersion))) {
+                $symfonyPackages += $composerJson['replace'];
```

Comment:
should use `?? []` to ensure we don't trigger a notice if the json does not contain this key (which is optional)

Reply:
`symfony/symfony` always has this "replace" key (same for the other ones that are unchecked btw.) I'd prefer not crippling the code for these checks...